### ISSUE TITLE
Compile error when importing macros

### ISSUE THREAD

**Issue by oblitzitate** [udayvir-singh/tangerine.nvim#23]
Just by importing macros via `(import-macros {: set!} :macros.opt-macros)`, I get the compile error below. Not sure the compiler should even evaluate the macro expressions given I haven't called a macro function just yet.

```
:: COMPILE ERROR
  xxx core/options.fnl
      Compile error in core/options.fnl:1:0
        Compile error in /home/oblitz/.config/nvim/fnl/macros/opt-macros.fnl:30:16
        Compile error in /home/oblitz/.config/nvim/fnl/macros/opt-macros.fnl:30:28
        expected quoted expression for expr

                      (quoted->fn value)
      stack traceback:
      	[C]: in function 'error'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:3510: in function 'assert-compile'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2242: in function '__fnl_global__assert_2dcompile'
      	/home/oblitz/.config/nvim/fnl/macros/lib/compile-macros.fnl:20: in function </home/oblitz/.config/nvim/fnl/macros/lib/compile-macros.fnl:18>
      	[C]: in function 'xpcall'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2754: in function 'macroexpand_2a'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2941: in function 'compile1'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:1247: in function 'compile_body'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:1258: in function 'special'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2779: in function 'compile1'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:3014: in function 'compile_top_target'
      	...
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2051: in function 'require-macros'
      	src/fennel/macros.fnl:343: in function <src/fennel/macros.fnl:328>
      	[C]: in function 'xpcall'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2754: in function 'macroexpand_2a'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:2941: in function 'compile1'
      	.../nvim/lazy/tangerine.nvim/lua/tangerine/fennel/1-2-1.lua:3208: in function 'string'
      	...e/nvim/lazy/tangerine.nvim/lua/tangerine/api/compile.lua:58: in function <...e/nvim/lazy/tangerine.nvim/lua/tangerine/api/compile.lua:48>
      	[C]: in function 'xpcall'
      	...e/nvim/lazy/tangerine.nvim/lua/tangerine/api/compile.lua:212: in function 'run'
      	[string ":lua"]:1: in main chunk
```

**Comment by udayvir-singh** [comment]
@oblitzitate Show me the contents of macro `macros.opt-macros`.

**Comment by oblitzitate** [comment]
opt-macros.fnl
```fennel
(local {: ->str : nil?} (require :macros.lib.types))
(local {: begins-with?} (require :macros.lib.string))
(import-macros {: quoted? 
                : fn?
                : quoted->fn
                : gensym-checksum
                : expand-exprs
                : vlua} :macros.lib.compile-macros)

(lambda set! [name ?value]
  "Set a vim option using the vim.opt.<name> API.
  Accepts the following arguments:
  name -> must be a symbol.
          - If it ends with '+' it appends to the current value.
          - If it ends with '-' it removes from the current value.
          - If it ends with '^' it prepends to the current value.
          - If it ends with '?' it ignores the value and returns the current
            value.
  value -> anything.
           - If it is not specified, whether the name begins with 'no' is used
             as a boolean value.
           - If it is a quoted expression or a function it becomes
             v:lua.<symbol>()."
  (assert-compile (sym? name) "expected symbol for name" name)
  (let [name (->str name)
        value (if (nil? ?value)
                  (not (begins-with? :no name))
                  ?value)
        value (if (quoted? value)
                (quoted->fn value)
                value)
        name (if (and (nil? ?value) (begins-with? :no name))
               (name:match "^no(.+)$")
               name)
        exprs (if (fn? value)
                [`(tset _G ,(->str (gensym-checksum value {:prefix "__"})) ,value)]
                [])
        value (if (fn? value)
                (vlua (gensym-checksum value {:prefix "__"}))
                value)
        exprs (doto exprs
                    (table.insert (match (name:sub -1)
                                    "+" `(: (. vim.opt ,(name:sub 1 -2)) :append ,value)
                                    "-" `(: (. vim.opt ,(name:sub 1 -2)) :remove ,value)
                                    "^" `(: (. vim.opt ,(name:sub 1 -2)) :prepend ,value)
                                    "?" `(: (. vim.opt ,(name:sub 1 -2)) :get)
                                    _ `(tset vim.opt ,name ,value))))]
    (expand-exprs exprs)))

(lambda local-set! [name ?value]
  "Set a vim option using the vim.opt_local.<name> API.
  Accepts the following arguments:
  name -> must be a symbol.
          - If it ends with '+' it appends to the current value.
          - If it ends with '-' it removes from the current value.
          - If it ends with with '^' it prepends to the current value.
          - If it ends with '?' it ignores the value and returns the current
            value.
  value -> anything.
           - If it is not specified, whether the name begins with 'no' is used
             as a boolean value.
           - If it is a quoted expression or a function it becomes
             v:lua.<symbol>()."
  (assert-compile (sym? name) "expected symbol for name" name)
  (let [name (->str name)
        value (if (nil? ?value)
                  (not (begins-with? :no name))
                  ?value)
        value (if (quoted? value)
                (quoted->fn value)
                value)
        name (if (and (nil? ?value) (begins-with? :no name))
               (name:match "^no(.+)$")
               name)
        exprs (if (fn? value)
                [`(tset _G ,(->str (gensym-checksum value {:prefix "__"})) ,value)]
                [])
        value (if (fn? value)
                (vlua (gensym-checksum value {:prefix "__"}))
                value)
        exprs (doto exprs
                    (table.insert (match (name:sub -1)
                                    "+" `(: (. vim.opt_local ,(name:sub 1 -2)) :append ,value)
                                    "-" `(: (. vim.opt_local ,(name:sub 1 -2)) :remove ,value)
                                    "^" `(: (. vim.opt_local ,(name:sub 1 -2)) :prepend ,value)
                                    "?" `(: (. vim.opt_local ,(name:sub 1 -2)) :get)
                                    _ `(tset vim.opt_local ,name ,value))))]
    (expand-exprs exprs)))

{: set!
 : local-set!}
```

compile-macros.fnl
```fennel
(local {: ->str} (require :macros.lib.types))
(local {: first
        : second} (require :macros.lib.seq))
(local {: djb2} (require :macros.lib.crypt))

(lambda expr->str [expr]
  `(string.gsub (macrodebug ,expr nil) "_%d+_auto" "#"))

(fn fn? [x]
  "Checks if `x` is a function definition.
  Cannot check if a symbol is a function in compile time."
  (and (list? x)
       (or (= `fn (first x))
           (= `hashfn (first x))
           (= `lambda (first x))
           (= `partial (first x)))))

(fn quoted? [x]
  "Check if `x` is a list that begins with `quote`."
  (and (list? x)
       (= `quote (first x))))

(lambda quoted->fn [expr]
  "Converts an expression like `(quote (+ 1 1))` into `(fn [] (+ 1 1))`."
  ; BUG:
  (assert-compile (quoted? expr) "expected quoted expression for expr" expr)
  (let [non-quoted (second expr)]
    `(fn [] ,non-quoted)))

(lambda quoted->str [expr]
  "Converts a quoted expression like `(quote (+ 1 1))` into an string with its shorthand form."
  (assert-compile (quoted? expr) "expected quoted expression for expr" expr)
  (let [non-quoted (second expr)]
    (.. "'" (view non-quoted))))

(lambda expand-exprs [exprs]
  "Converts a list of expressions into either an expression - if only one
  expression is in the list - or a do-expression containing the expressions."
  (if (> (length exprs) 1)
    `(do
       ,(unpack exprs))
    (first exprs)))

(lambda gensym-checksum [x ?options]
  "Generates a new symbol from the checksum of the object passed as a parameter
  after it is casted into an string using the `view` function.
  You can also pass a prefix or a suffix into the options optional table.
  This function depends on the djb2 hash function."
  (let [options (or ?options {})
        prefix (or options.prefix "")
        suffix (or options.suffix "")]
    (sym (.. prefix (djb2 (view x)) suffix))))

(lambda vlua [x]
  "Return a symbol mapped to `v:lua.%s()` where `%s` is the symbol."
  (assert-compile (sym? x) "expected symbol for x" x)
  (string.format "v:lua.%s()" (->str x)))

{: expr->str
 : fn?
 : quoted?
 : quoted->fn
 : quoted->str
 : expand-exprs
 : gensym-checksum
 : vlua}
```

**Comment by udayvir-singh** [comment]
@oblitzitate This is not an error with tangerine. You simply are using macro's wrong.

Here an example to illustrate your problem:

```
;; foo.fnl

(lambda double [x]
  '(+ x x))

{: double}
```

```
;; baz.fnl

(require-macros :foo)

(lambda example [x]
  ...
  :return
  (double x)) ;; ERROR
```

The macro `example` would throw an error as the macro will expand **inside the body of the function** rather than expanding outside.

So our macro `example` would become:
```
(lambda example [x]
  ...
  :return
  (+ x x)) ; Hence the error, since you are trying to add two symbols rather the returning a quoted value.
```

TLDR: Don't try to return macro expansions from the body of a macro.

**Comment by oblitzitate** [comment]
If you're referring to `(quoted->fn value)`, I don't think the problem has to do with what you mentioned. 

The error seems to be coming from passing an invalid argument into the `quoted->fn` function, but I haven't actually called `set!` yet so it shouldn't pass anything to `quoted->fn` function just yet.

The code is a copy-paste of https://github.com/datwaft/themis.nvim/blob/main/fnl/themis/opt.fnl which is based on https://github.com/nyoom-engineering/nyoom.nvim/blob/main/fnl/macros.fnl. Their code seems to work fine using `hotpot`.

**Comment by udayvir-singh** [comment]
I just read @datwaft config, and the only difference I could find b/w his and your config is that he required `quoted->fn` normally as a function as opposed to a macro.

https://github.com/datwaft/themis.nvim/blob/6f86b1436aebeb8344237f9e937f090ae302c824/fnl/themis/opt.fnl#L3

```fennel
(local {: quoted? : fn?
        : quoted->fn
        : gensym-checksum
        : expand-exprs
        : vlua} (require :themis.lib.compile-time))
```

**Comment by udayvir-singh** [comment]
Other than that I have no clue on what might be causing your error as macros are mostly hard to debug.

You can also just use [hibiscus](https://github.com/udayvir-singh/hibiscus.nvim) macro library if you want some prebuilt macros.

**Comment by oblitzitate** [comment]
Thanks for trying to help. Yeah, I just wanted to fiddle around with macros just to learn, but I'm aware hibiscus exists. Will probably end up using it.

**Comment by udayvir-singh** [comment]
@oblitzitate If you are trying to learn how to write macros, just write your own from scratch and don't try copy other people's macros, as you might just end up wasting a lot of your time trying to figure out the spaghetti macro code.
